From 4e2aad26c10f786745f2e75f9b4034878f7502ff Mon Sep 17 00:00:00 2001
From: Jiri Benc <jbenc@redhat.com>
Date: Fri, 12 Sep 2014 14:30:07 -0400
Subject: [net] openvswitch: Fix a double free bug for the sample action

Message-id: <16f331ad52ce60756e8b98c4c3cca8249ceefdf4.1410531783.git.jbenc@redhat.com>
Patchwork-id: 94037
O-Subject: [RHEL7.1 net PATCH 34/50] openvswitch: Fix a double free bug for the sample action
Bugzilla: 1110384
RH-Acked-by: Flavio Leitner <fbl@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110384
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7948291

commit fe984c08e20f0fc2b4666bf8eeeb02605568387b
Author: Andy Zhou <azhou@nicira.com>
Date:   Tue May 6 17:23:48 2014 -0700

    openvswitch: Fix a double free bug for the sample action

    When sample action returns with an error, the skb has already been
    freed. This patch fix a bug to make sure we don't free it again.
    This bug introduced by commit ccb1352e76cff05 (net: Add Open vSwitch
    kernel components.)

    Signed-off-by: Andy Zhou <azhou@nicira.com>
    Signed-off-by: Pravin B Shelar <pshelar@nicira.com>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/openvswitch/actions.c b/net/openvswitch/actions.c
index b58595b..6054554 100644
--- a/net/openvswitch/actions.c
+++ b/net/openvswitch/actions.c
@@ -556,6 +556,8 @@ static int do_execute_actions(struct datapath *dp, struct sk_buff *skb,
 
   case OVS_ACTION_ATTR_SAMPLE:
    err = sample(dp, skb, a);
+   if (unlikely(err)) /* skb already freed. */
+    return err;
    break;
   }
 
-- 
1.7.1